### PR TITLE
Fix deadlock where opening database fails 

### PR DESCRIPTION
**Pull request by padelt**
...without setting self.exception; avoid deadlock after race condition where command is enqueued before thread can signal an exception #90

### CODE REVIEW

**Comment by ronnymajani** [comment]
@mpenkov Any updates on getting this PR merged?

I've tested this PR in our own fork and it pretty much fixes the issue, and has been stable so far (been using this patch for a couple of months now).

Just though I'd ping you on this, as it would be useful patch for others to have
And thanks for this module, it's sweet, and easily extendable ❤️

File: sqlitedict.py [hunk 1]
```diff
@@ -390,19 +395,34 @@ def __init__(self, filename, autocommit, journal_mode):
...
-        conn.commit()
-        cursor.execute('PRAGMA synchronous=OFF')
+        try:
+            if self.autocommit:
+                conn = sqlite3.connect(self.filename, isolation_level=None, check_same_thread=False)
+            else:
+                conn = sqlite3.connect(self.filename, check_same_thread=False)
+        except Exception as ex:
+            self.log.exception("Failed to initialize connection for filename: %s" % self.filename)
+            self.exception = (e_type, e_value, e_tb) = sys.exc_info()
```
Comment (mpenkov) [inline sqlitedict.py:411]: What's the point of declaring the new variables?

```suggestion
            self.exception = sys.exc_info()
```

File: sqlitedict.py [hunk 2]
```diff
@@ -552,6 +573,28 @@ def close(self, force=False):
             self.select_one('--close--')
             self.join()
 
+    def wait_for_initialization(self):
```
Comment (mpenkov) [inline sqlitedict.py:576]: I think it's worth making this thing explicitly internal, so people don't touch it (or at least expect bad things to happen if they do).

```suggestion
    def _wait_for_initialization(self):
```

File: sqlitedict.py [hunk 3]
```diff
@@ -390,19 +395,34 @@ def __init__(self, filename, autocommit, journal_mode):
         self.reqs = Queue()
         self.setDaemon(True)  # python2.5-compatible
         self.exception = None
+        self.initialized = None
```
Comment (mpenkov) [inline sqlitedict.py:398]:
```suggestion
        self._initialized = None
```

**Comment by mpenkov** [comment]
Thank you for reminding me about this. Let's work on getting this merged. I left you some minor comments, please have a look.

**Comment by ronnymajani** [comment]
This isn't my PR, so I'll wait for the original author to jump in @padelt 
But if they don't then I'll fork this and make a copy PR to get this merged in faster

**Comment by padelt** [comment]
Hey everyone - nearly forget about this PR.
@mpenkov Thanks for the comments. I just added two commits which should address those.
@ronnymajani Thank you for deliberately not stepping on toes. ;-) I don't mind whose PR gets merged.
Whatever is easier for you all!

**Comment by padelt** [comment]
Had to rename the variable again to avoid messing up `threading.Thread`'s internal state.

@mpenkov Please have a look [at the line](https://github.com/RaRe-Technologies/sqlitedict/pull/107/files#diff-a53c7bf78c0f4fca3d6533a0303633a5c71a8fb697edc05322bdfe9ae550c175R599) `# endclass SqliteMultithread` - I inserted a space to please Flake8 - but I have no idea what that comment is supposed to do - so maybe it was important to have that exact syntax?

Tests now pass: https://github.com/padelt/sqlitedict/runs/1757442319

**Comment by mpenkov** [comment]
Merged. Thank you for your contribution and your patience @padelt !

**Comment by shawnboltz** [comment]
Hello, are there any plans to release this in a new version to PyPI? It appears to address an issue that I'm having with the latest official release. Thanks!

**Pull request merged**